_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. For example, it is not permitted for one ticket to be standing while the other for the normal seated area.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class NormalSeat(row: Int, column: Int) extends Seat
+case class VipTribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />   
+
+We also have:  
+
+```scala
+case class Ticket[SeatType <: Seat](
+  verificationCode: String,
+  price: Int,
+  seat: SeatType
+)
+```
+
+```scala
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribune("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribune</code>. The compiler is satisfied since this is a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> on line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
```

Comment:
I'd replace 
> From the user ticket on line 2 it thinks it is <code>VipTribune</code>.

with `Regarding user ticket on line 2, it thinks it is <code>VipTribune</code>.` 

Also here
> Bad news since type <code>SeatType</code> on line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  

I'd would rephrase the last part to `on line 3 refer to, well, the same type! ` -> moved the "to" so it's before the "well".

Reply:
**First point**
I didn't understand you wanted a consequential effect. I'd still remove the "the" before "user ticket" and add a comma before "it thinks":
> From user ticket on line 2, it thinks it is VipTribune.

**Second point**
I didn't find any grammatical rule about that, but orally I think it sounds better the way I suggested. Still, this isn't an issue for me, just a suggestion.